src/GAP.jl

```diff
@@ -419,6 +420,37 @@ function prompt()
     reset_GAP_ERROR_OUTPUT()
 end
 
+"""
+    GAP.randseed!([seed::Integer])
+
+Reseed GAP's global RNG with `seed`.
+
+The given `seed` must be a non-negative integer.
+When `seed` is not specified, a random seed is generated from Julia's global RNG.
+
+For a fixed seed, the stream of generated numbers is allowed to change between
+different versions of GAP.
+"""
+function randseed!(seed::Union{Integer,Nothing}=nothing)
+    if seed === nothing
+        seed = rand(UInt128)
+    end
+    if 0 <= seed < typemax(Int) ÷ 2
+        # avoid too small seeds by hashing them
+        seed = hash(seed, 0x886b5530f00038ef % UInt)
+    end
+    # Although GAP's `Reset` method can accept directly integers (also negative
+    # ones), we harmonize here the behavior with what Julia does, i.e. for two
+    # integers with the same value, even of different types, the initialization
+    # is the same (e.g. with `1` or `0x1`); so we re-use Julia's
+    # `Random.make_seed` internal function (which errors on negative argument)
+    # to normalize this, and create a string out of it.
+    str = String(reinterpret(UInt8, Random.make_seed(seed)))
```

Comment:
I don't mind the use of `Random.make_seed` but I don't understand why it is needed to "harmonize" anything -- why can't one just use e.g. `seed % UInt` ? That "harmonizes" the integer, too, doesn't it?

Reply:
Doing `seed % UInt` limits the number of possible random sequences to 2^64, when the RNG's state can handle much more. Also, I want to enforce getting two different streams when two different seeds are given. With `seed % UInt`, 1 and `1 + big(2)^64` would lead to the same sequence. So rather than accepting only seeds in the [0, 2^64-1] interval to avoid this problem, we can just accept any seed >= 0 using `Random.make_seed`.

Reply:
> "harmonizes" 

BTW, do you suggest a better term than "harmonize"? :)

Reply:
> I want to enforce getting two different streams when two different seeds are given 

But then, I can't do that for `GlobalRandomSource`, as it takes the seed module 2^28 :shrug:

Reply:
I am still confused by how this is phrased. I am actually fine with the word "harmoinzed"; what confuses me why this is an issue that involves `make_seed` in the first place. So yeah, `seed % UInt` truncates values to a finite space; but anyway, all RNGs will use a finite state, including Mersenne twister. But yeah, it cuts too much; so you could use `convert(GapObj, seed)` or perhaps `convert(GapObj, big(seed))` or the same but using `julia_to_gap`, to get a GAP integer, and then pass that on. This would also ensure that 1 and 0x1 are treated the same *and* still allow negative integers.

I am not saying you *should* do it that way; there may be good reasons why  using `Random.make_seed` is superior! Rather, my point is that the argument given in the comment does not fit the conclusion, at least in my mind.

Reply:
> the argument given in the comment does not fit the conclusion, at least in my mind.

It was just a conclusion in my mind :) It's because I'm fairly familiar with `Random.make_seed`, which transforms an integer into the native seed format of dSFMT (a `Vector{UInt32}`). My reasoning was that constructing in-place a `String` from this array and feeding it to GAP's `Init` method would be optimal. But I had two optimizations bugs:
1) `String(UInt8[...])` constructs the string by re-using the array storage, but here I was using `String(reinterpret(UInt8, UInt32[...]))`, in which case the content of the array is copied;
2) much more important, passing a Julia `String` to `Globals.Reset` results in the argument not being considered as a string on the GAP side (and thought a GAP string is created from that argument, I'm not sure in which way), the solution is of course to call `julia_to_gap(str)`, which improves greatly performance. 

Anyway, in the end using `Random.make_seed` or `convert(GapObj, seed)` actually results in the same performance, so I'm happy with any. I guess you prefer the latter, so I can change the PR to using that.

Reply:
No, I honestly have no preference between `Random.make_seed` or `convert(GapObj, seed)` -- but I do want to be able to understand the final code, if possible, and reading the comment and code it felt to me as if something was missing -- either in my understanding, or in the code/comment, or both. :-)

Reply:
So I updated with passing directly the integer seed to GAP's seeding routine. It's more straightforward, and actually allows to get the same streams than in GAP for a given seed, which might be a feature. The comment containing "harmonize" is thus deleted :)

Reply:
I don't know if there will be interest for #544, but having it this way also wouldn't create discrepancy between `randseed!(seed)` and `Random.seed!(GAP.GlobalMersenneTwister, seed)` or `Random.seed!(GAP.default_rng(), seed)` (whatever we call it).